iceoryx-sys/src/sample.rs

```diff
@@ -0,0 +1,218 @@
+    ///
+    /// The `len` argument is the number of **elements**, not the number of bytes.
+    ///
+    /// This function is safe, but dereferencing the return value is unsafe.
+    /// See the documentation of [`slice::from_raw_parts`] for slice safety requirements.
+    #[must_use]
+    #[inline]
+    pub fn slice_from_raw_parts(sample: RawSample<T>, len: usize) -> RawSample<[T]> {
+        // SAFETY: `self.as_payload_ptr` returns a non-null ptr
+        unsafe { Self::new_unchecked(slice::from_raw_parts(sample.as_payload_ptr(), len)) }
+    }
+
+    /// Returns the length of a non-null raw slice.
+    #[must_use]
+    #[inline]
+    pub fn len(self) -> usize {
+        // SAFETY: `self.as_payload_ptr` returns a non-null ptr
+        unsafe { (*self.as_payload_ptr()).len() }
+    }
+}
+
+impl<T: ?Sized> Clone for RawSample<T> {
+    #[inline(always)]
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<T: ?Sized> Copy for RawSample<T> {}
+
+impl<T: ?Sized> fmt::Debug for RawSample<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Pointer::fmt(&self.as_payload_ptr(), f)
+    }
+}
+
+impl<T: ?Sized> fmt::Pointer for RawSample<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Pointer::fmt(&self.as_payload_ptr(), f)
+    }
+}
+
+/// A `*mut T` non-zero sample pointer to the user payload.
+#[repr(transparent)]
+pub struct RawSampleMut<T: ?Sized> {
+    payload: *mut T,
+}
+
+impl<T: ?Sized> RawSampleMut<T> {
+    /// Creates a new `RawSampleMut`.
+    ///
+    /// # Safety
+    ///
+    /// `payload` must be non-null.
+    #[inline]
+    pub unsafe fn new_unchecked(payload: *mut T) -> Self {
+        debug_assert!(
+            !payload.is_null(),
+            "RawSampleMut::new_unchecked requires that the payload pointer is non-null"
+        );
+        Self { payload }
+    }
+
+    /// Creates a new `RawSampleMut`.
+    #[allow(clippy::not_unsafe_ptr_arg_deref)] // false positive
+    #[inline]
+    pub fn new(payload: *mut T) -> Option<Self> {
+        if !payload.is_null() {
+            // SAFETY: `payload` pointer is checked to be non-null
+            Some(unsafe { Self::new_unchecked(payload) })
+        } else {
+            None
+        }
+    }
+
+    /// Casts to a `RawSampleMut` of another type.
+    #[must_use]
+    #[inline]
+    pub fn cast<U>(self) -> RawSampleMut<U> {
+        // SAFETY: `self.as_payload_mut_ptr` returns a non-null ptr
+        unsafe { RawSampleMut::new_unchecked(self.as_payload_mut_ptr().cast::<U>()) }
+    }
+
+    /// Acquires the underlying payload pointer as `*const` pointer.
+    #[must_use]
+    #[inline(always)]
+    pub fn as_payload_ptr(self) -> *const T {
+        self.as_payload_mut_ptr()
+    }
+
+    /// Acquires the underlying payload pointer as `*mut` pointer.
+    #[must_use]
+    #[inline(always)]
+    pub fn as_payload_mut_ptr(self) -> *mut T {
```

Comment:
Shouldn't it be `mut self` here?

Reply:
Hehe, in theory yes but I took inspiration from [NonNull::as_ptr](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_ptr). Initially I tried to use `NonNull` instead of an own type but then I discovered that I could not get a `*const T` out of it and finally replicated `NonNull` into these two structs

Reply:
The counsel with the pillow helped. `mut self` would even cause a warning. The method does no borrow but takes the ownership of the sample. The question is now why not using a reference? My guess is that taking the ownership and invoking a copy is more performant than using a reference. The struct has the the same size as a raw pointer and therefore passing by value or by reference has the same cost. Using the reference does add an indirection which is unnecessary in this case. It should not be a big penalty but I would keep the design similar to `NonNull`